IllegalAccessError occurs when GLIBC defines a proxy class for a package private bean class in Spring

**Describe the bug**  
If a Spring application uses some aspect of Spring that requires the use of a proxy (e.g. `@Async` or `@Transactional`) then Spring will end up using CLIBC if it needs to proxy the class of the bean.  If the bean class is package private then something like the following exception will occur when running on Liberty:

```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'packagePrivateBean' defined in URL [jar:file:/home/tjwatson/dev/liberty/source/open-liberty/dev/build.image/wlp/usr/servers/SpringBootTests/workarea/spring/spring.thin.apps/io.openliberty.springboot.fat30.app-0.0.1-SNAPSHOT.spring!/BOOT-INF/classes/com/ibm/ws/springboot/fat30/test/app/PackagePrivateBean.class]: Could not generate CGLIB subclass of class com.ibm.ws.springboot.fat30.test.app.PackagePrivateBean: Common causes of this problem include using a final class or a non-visible class
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:607) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1448) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1358) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.resolveFieldValue(AutowiredAnnotationBeanPostProcessor.java:785) ~[spring-beans-6.1.20.jar:6.1.20]
	... 27 common frames omitted
Caused by: org.springframework.aop.framework.AopConfigException: Could not generate CGLIB subclass of class com.ibm.ws.springboot.fat30.test.app.PackagePrivateBean: Common causes of this problem include using a final class or a non-visible class
	at org.springframework.aop.framework.CglibAopProxy.buildProxy(CglibAopProxy.java:231) ~[spring-aop-6.1.20.jar:6.1.20]
	at org.springframework.aop.framework.CglibAopProxy.getProxy(CglibAopProxy.java:164) ~[spring-aop-6.1.20.jar:6.1.20]
	at org.springframework.aop.framework.ProxyFactory.getProxy(ProxyFactory.java:110) ~[spring-aop-6.1.20.jar:6.1.20]
	at org.springframework.aop.framework.AbstractAdvisingBeanPostProcessor.postProcessAfterInitialization(AbstractAdvisingBeanPostProcessor.java:127) ~[spring-aop-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsAfterInitialization(AbstractAutowireCapableBeanFactory.java:438) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1809) ~[spring-beans-6.1.20.jar:6.1.20]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600) ~[spring-beans-6.1.20.jar:6.1.20]
	... 36 common frames omitted
Caused by: org.springframework.cglib.core.CodeGenerationException: java.lang.IllegalAccessError-->com/ibm/ws/springboot/fat30/test/app/PackagePrivateBean is not visible
	at org.springframework.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:521) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:375) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.proxy.Enhancer.generate(Enhancer.java:575) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData.lambda$new$1(AbstractClassGenerator.java:107) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.core.internal.LoadingCache.lambda$createEntry$1(LoadingCache.java:52) ~[spring-core-6.1.20.jar:6.1.20]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[na:na]
	at org.springframework.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:57) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.core.internal.LoadingCache.get(LoadingCache.java:34) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:129) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:321) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.proxy.Enhancer.createHelper(Enhancer.java:562) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.cglib.proxy.Enhancer.createClass(Enhancer.java:407) ~[spring-core-6.1.20.jar:6.1.20]
	at org.springframework.aop.framework.ObjenesisCglibAopProxy.createProxyClassAndInstance(ObjenesisCglibAopProxy.java:62) ~[spring-aop-6.1.20.jar:6.1.20]
	at org.springframework.aop.framework.CglibAopProxy.buildProxy(CglibAopProxy.java:222) ~[spring-aop-6.1.20.jar:6.1.20]
	... 42 common frames omitted
Caused by: java.lang.IllegalAccessError: com/ibm/ws/springboot/fat30/test/app/PackagePrivateBean is not visible
	at java.base/java.lang.ClassLoader.defineClassImpl(Native Method) ~[na:na]
	at java.base/java.lang.ClassLoader.defineClassInternal(ClassLoader.java:474) ~[na:na]
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:435) ~[na:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:586) ~[na:na]
	at org.springframework.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:518) ~[spring-core-6.1.20.jar:6.1.20]
	... 55 common frames omitted
```

This is because Spring uses the thread context class loader to define the GLIBC proxy class which extends the package private bean class.  To do that the proxy class must be in the same class as the package private bean class.  The bean class is defined by the application class loader while the GLIBC proxy class gets defined by the thread context class loader.  In liberty the thread context class loader is different from the application class loader.  The implications of this is that the GLIBC proxy class cannot have visibility to the package private bean class because package private access requires the two classes to be defined by the same class loader.

**Steps to Reproduce**  
See above

**Expected behavior**  
No exceptions to occur when GLIBC defines a proxy class for a package private bean class.

**Diagnostic information:**  
 - OpenLiberty Version: all versions up and including 25.0.0.5
 - Affected feature(s): all
 - Java Version: all

--- a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
+++ b/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
@@ -847,7 +847,8 @@ public class AppClassLoader extends ContainerClassLoader implements SpringLoader
         return sb.toString();
     }
 
-    Class<?> defineClassForTCCL(String name, byte[] b, ProtectionDomain protectionDomain) {
+    @Override
+    public Class<?> publicDefineClass(String name, byte[] b, ProtectionDomain protectionDomain) {
         return defineClass(name, b, 0, b.length, protectionDomain);
     }
 }

--- a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
+++ b/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/AppClassLoader.java
@@ -846,4 +846,8 @@ public class AppClassLoader extends ContainerClassLoader implements SpringLoader
 
         return sb.toString();
     }
+
+    Class<?> defineClassForTCCL(String name, byte[] b, ProtectionDomain protectionDomain) {
+        return defineClass(name, b, 0, b.length, protectionDomain);
+    }
 }

--- a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/SpringLoader.java
+++ b/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/SpringLoader.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2014 IBM Corporation and others.
+ * Copyright (c) 2014, 2025 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -13,6 +13,7 @@
 package com.ibm.ws.classloading.internal;
 
 import java.lang.instrument.ClassFileTransformer;
+import java.security.ProtectionDomain;
 
 /**
  * Declare the methods expected by Spring's ReflectiveLoadTimeWeaver.
@@ -23,4 +24,7 @@ interface SpringLoader {
     boolean addTransformer(ClassFileTransformer cft);
 
     ClassLoader getThrowawayClassLoader();
+
+    // This is needed by org.springframework.cglib.core.ReflectUtils which will reflectively call this if available
+    Class<?> publicDefineClass(String name, byte[] b, ProtectionDomain protectionDomain);
 }

--- a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ThreadContextClassLoader.java
+++ b/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ThreadContextClassLoader.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2013, 2022 IBM Corporation and others.
+ * Copyright (c) 2013, 2025 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -16,6 +16,7 @@ import java.io.IOException;
 import java.net.URL;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
+import java.security.ProtectionDomain;
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.LinkedHashSet;
@@ -215,4 +216,11 @@ public class ThreadContextClassLoader extends UnifiedClassLoader implements Keye
         return classLoader == appLoader;
     }
 
+    // This is needed by spring to ensure GLIBC generated classes are defined by the app classloader
+    public Class<?> publicDefineClass(String name, byte[] b, ProtectionDomain protectionDomain) {
+        if (appLoader instanceof AppClassLoader) {
+            return ((AppClassLoader) appLoader).defineClassForTCCL(name, b, protectionDomain);
+        }
+        return defineClass(name, b, 0, b.length, protectionDomain);
+    }
 }

--- a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ThreadContextClassLoader.java
+++ b/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ThreadContextClassLoader.java
@@ -216,10 +216,13 @@ public class ThreadContextClassLoader extends UnifiedClassLoader implements Keye
         return classLoader == appLoader;
     }
 
-    // This is needed by spring to ensure GLIBC generated classes are defined by the app classloader
+    // This is reflectively called by Spring for GLIBC proxy classes.
+    // To work properly with package private application class beans these proxies
+    // must be defined with the app loader so they have proper visibility to their
+    // super classes.
     public Class<?> publicDefineClass(String name, byte[] b, ProtectionDomain protectionDomain) {
-        if (appLoader instanceof AppClassLoader) {
-            return ((AppClassLoader) appLoader).defineClassForTCCL(name, b, protectionDomain);
+        if (appLoader instanceof SpringLoader) {
+            return ((SpringLoader) appLoader).publicDefineClass(name, b, protectionDomain);
         }
         return defineClass(name, b, 0, b.length, protectionDomain);
     }

--- a/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java
+++ b/dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/UnifiedClassLoader.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2022 IBM Corporation and others.
+ * Copyright (c) 2011, 2025 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -17,6 +17,7 @@ import java.lang.instrument.ClassFileTransformer;
 import java.net.URL;
 import java.security.AccessController;
 import java.security.PrivilegedActionException;
+import java.security.ProtectionDomain;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.EnumSet;
@@ -279,4 +280,10 @@ public class UnifiedClassLoader extends LibertyLoader implements SpringLoader {
     public Bundle getBundle() {
         return null;
     }
+
+    @Override
+    public Class<?> publicDefineClass(String name, byte[] b, ProtectionDomain protectionDomain) {
+        return defineClass(name, b, 0, b.length, protectionDomain);
+    }
+
 }

--- a/dev/io.openliberty.springboot.fat30.app/build.gradle
+++ b/dev/io.openliberty.springboot.fat30.app/build.gradle
@@ -13,10 +13,9 @@
 
 // https://plugins.gradle.org/plugin/org.springframework.boot
 plugins {
-  id 'org.springframework.boot' version '3.0.7'
+  id 'org.springframework.boot' version '3.3.12'
 }
 apply from: '../wlp-gradle/subprojects/spring.gradle'
-def sv = springVersions[ '3.0' ]
 
 apply plugin: 'java'
 apply plugin: 'io.spring.dependency-management'
@@ -28,22 +27,22 @@ sourceCompatibility = 17
 apply from: '../wlp-gradle/subprojects/maven-central-mirror.gradle'
 
 dependencies {
-  implementation('org.springframework.boot:spring-boot' + ':' + sv['springBoot.fat'])
-  implementation('org.springframework.boot:spring-boot-autoconfigure' + ':' + sv['springBoot.fat'])
+  implementation('org.springframework.boot:spring-boot')
+  implementation('org.springframework.boot:spring-boot-autoconfigure')
 
-  implementation('org.springframework.boot:spring-boot-starter-web') // + ':' + sv['springBoot']) // x
-  implementation('org.springframework.boot:spring-boot-starter-thymeleaf') // + ':' + sv['springBoot']) // x
-  implementation('org.springframework.boot:spring-boot-starter-security') // + ':' + sv['springBoot']) // x
+  implementation('org.springframework.boot:spring-boot-starter-web')
+  implementation('org.springframework.boot:spring-boot-starter-thymeleaf')
+  implementation('org.springframework.boot:spring-boot-starter-security')
 
-  implementation('org.springframework:spring-beans' + ':' + sv['spring.fat'])
-  implementation('org.springframework:spring-core' + ':' + sv['spring.fat'])
-  implementation('org.springframework:spring-context' + ':' + sv['spring.fat'])
-  implementation('org.springframework:spring-web' + ':' + sv['spring.fat'])
-  implementation('org.springframework:spring-webmvc' + ':' + sv['spring.fat'])
+  implementation('org.springframework:spring-beans')
+  implementation('org.springframework:spring-core')
+  implementation('org.springframework:spring-context')
+  implementation('org.springframework:spring-web')
+  implementation('org.springframework:spring-webmvc')
 
-  implementation('org.springframework.security:spring-security-config' + ':' + sv['springSecurity.fat'])
-  implementation('org.springframework.security:spring-security-core' + ':' + sv['springSecurity.fat'])
-  implementation('org.springframework.security:spring-security-web' + ':' + sv['springSecurity.fat'])
+  implementation('org.springframework.security:spring-security-config')
+  implementation('org.springframework.security:spring-security-core')
+  implementation('org.springframework.security:spring-security-web')
 
   implementation('org.webjars:bootstrap:5.2.3')
   implementation('org.webjars:webjars-locator:0.46')

--- a/dev/io.openliberty.springboot.fat30.app/src/main/java/com/ibm/ws/springboot/fat30/test/app/PackagePrivateBean.java
+++ b/dev/io.openliberty.springboot.fat30.app/src/main/java/com/ibm/ws/springboot/fat30/test/app/PackagePrivateBean.java
@@ -1,0 +1,26 @@
+/*******************************************************************************
+ * Copyright (c) 2025 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.springboot.fat30.test.app;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Component;
+
+@Component
+class PackagePrivateBean {
+	@Async
+	CompletableFuture<String> testPackagePrivate() {
+		return CompletableFuture.completedFuture("PASSED");
+	}
+}

--- a/dev/io.openliberty.springboot.fat30.app/src/main/java/com/ibm/ws/springboot/fat30/test/app/TestApplication.java
+++ b/dev/io.openliberty.springboot.fat30.app/src/main/java/com/ibm/ws/springboot/fat30/test/app/TestApplication.java
@@ -14,16 +14,20 @@ package com.ibm.ws.springboot.fat30.test.app;
 
 import jakarta.servlet.ServletContext;
 
+import java.util.concurrent.ExecutionException;
+
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.core.env.Environment;
+import org.springframework.scheduling.annotation.EnableAsync;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
 
 @SpringBootApplication
 @RestController
+@EnableAsync
 public class TestApplication {
 	public static final String TEST_ATTR = "test.weblistener.attr";
 	@Autowired
@@ -31,6 +35,9 @@ public class TestApplication {
 	
 	@Autowired
 	private Environment env;
+
+	@Autowired
+	private PackagePrivateBean packagePrivateBean;
 
 	public static void main(String[] args) {
 		SpringApplication.run(TestApplication.class, args);
@@ -81,7 +88,12 @@ public class TestApplication {
 	public void throwFileSystemNotFoundException() {
 		throw new java.nio.file.FileSystemNotFoundException("Thrown on purpose for FAT test. Default error page.");
 	}
-	
+
+	@RequestMapping("/testPackagePrivateBean")
+	public String testPackagePrivateBean() throws InterruptedException, ExecutionException {
+		return packagePrivateBean.testPackagePrivate().get();
+	}
+
 	static final String 
 	IbmApiClazzName = "com.ibm.websphere.application.ApplicationMBean",
 	TpClazzName 	= "jakarta.mail.Message";

--- a/dev/io.openliberty.springboot.fat30_fat/fat/src/com/ibm/ws/springboot/support/fat/CommonWebServerTests30.java
+++ b/dev/io.openliberty.springboot.fat30_fat/fat/src/com/ibm/ws/springboot/support/fat/CommonWebServerTests30.java
@@ -14,6 +14,8 @@ package com.ibm.ws.springboot.support.fat;
 
 import static componenttest.custom.junit.runner.Mode.TestMode.FULL;
 
+import java.io.IOException;
+
 import org.junit.After;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -21,6 +23,7 @@ import org.junit.runner.RunWith;
 import componenttest.annotation.MinimumJavaLevel;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
+import componenttest.topology.utils.HttpUtils;
 
 @RunWith(FATRunner.class)
 @Mode(FULL)
@@ -54,5 +57,10 @@ public class CommonWebServerTests30 extends CommonWebServerTests {
         // A variation of 'testBasicSpringBootApplication30Servlet60'.
         // The different behavior is triggered by the test name.
         testBasicSpringBootApplication();
+    }
+
+    @Test
+    public void testPackagePrivateBean() throws IOException {
+        HttpUtils.findStringInUrl(server, "/testPackagePrivateBean", "PASSED");
     }
 }

